### PR TITLE
refactor(scout-for-lol): add split-capable runtime roles

### PR DESCRIPTION
**Pull request by shepherdjerred**
## Why

One process boots everything, gated by dev-only env guards, so the runtime cannot be split into the application/gateway/activity-worker topology the durable-architecture program deploys in Wave 6 — and the gateway-before-HTTP ordering constraint dissolved when #2819 removed the web surface's gateway-cache reads.

**Stacked on #2819.**

## What

- **A closed `ScoutRuntimeRole` vocabulary** (combined | application | gateway | activity-worker) selected by `SCOUT_RUNTIME_ROLE` (default combined), expressed as a **capability table in a zod-only leaf module**: which subsystems each role starts is a value that can be read, printed, and asserted — a role cannot half-configure a pod by forgetting a branch, because there are no branches to forget. `src/runtime/` (plan/boot/subsystems) acts on the table, replacing `startup.ts`.
- The dev-only `ENABLE_DISCORD_GATEWAY`/`ENABLE_BACKGROUND_JOBS` guards are retired in favor of roles; dev tooling migrated.
- **application**: HTTP/tRPC/OAuth/SSE served immediately (no shard wait), interactive+lake Temporal workers, DuckDB report lake, and sole ownership of the DB-sweeping metric collectors (`metrics/sweep-policy.ts` — other roles serve process/worker metrics only).
- **gateway**: the shard, commands, guild lifecycle, and the gateway-coupled Hey Scout voice assistant (explicitly stateful); voice drains before Temporal on shutdown; no Temporal workers — commands start workflows they do not execute.
- **activity-worker**: realtime/background/competition workers with REST-tokened Discord delivery — `discord/rest-identity.ts` gives non-gateway roles a construction-time REST identity, closing the residual `client.rest` coupling #2819 flagged.
- gateway and activity-worker expose health/metrics via a minimal `http/admin-server.ts`, not the product HTTP surface. Same-image/different-args per the workflow-worker precedent; roles documented in the backend README; `runtime/` registered in the architecture layers.
- Kubernetes still boots `combined` — the deployment split is Wave 6's change, and this PR is its rollback floor.

## Verification

- `bunx turbo run typecheck test lint --filter=@scout-for-lol/backend --filter=scout-for-lol` — 26/26 tasks, 392 test files green: exhaustive role-capability table test, per-role boot tests with heavy subsystems mocked, combined-equivalence assertions, invalid-role rejection; architecture 1218+ modules clean; `bun run jscpd` green after consolidating a duplicated test fixture.
- **Not run**: real per-role pods (no deployment exists until Wave 6); live gateway/voice boot on a role pod.

Part of SJ-189 (Wave 2 of the Scout durable-architecture program, SJ-186).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_013qRdG6fzJr3THzYzAJwTZB

### CODE REVIEW

**Comment by shepherdjerred** [comment]
This change is part of the following stack:

- #2826 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

**Comment by shepherdjerred** [comment]
@codex review

<sub>Automated re-review request for `783ef85` from the CI review gate.</sub>

**Comment by shepherdjerred** [comment]
@codex review

<sub>Automated re-review request for `45a476e` from the CI review gate.</sub>

**Comment by shepherdjerred** [comment]
@codex review

<sub>Automated re-review request for `7099273` from the CI review gate.</sub>

File: packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    databaseMetricSweeps: false,
+    databaseSeeding: false,
+  },
+  "activity-worker": {
+    championAssets: true,
+    voiceAssistant: false,
+    reportLakeFold: false,
+    temporalWorkers: DISCORD_WORKERS,
+    deferredTemporalWorkers: NO_WORKERS,
+    discordGateway: false,
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts:233]: Fixed, and the destructive path no longer takes a client at all.

`runDataValidation`'s guild half now resolves orphans through the merged install port instead of `client.guilds.cache` (`league/tasks/cleanup/validate-data.ts`). Two steps: `installedGuildIdsAmong` answers for every stored guild in one query and a live row is proof of presence; only the remaining deletion candidates are confirmed against Discord with `isScoutInstalledInGuild`, which throws rather than returning `false` when it cannot reach Discord — so "Scout could not ask" can never become "Scout was removed" on a path wired to `deleteMany`. This is correct for every role including `combined`, where a shard that had not finished backfilling produced the same class of wrong answer on a smaller scale.

While in there I fixed the same failure mode in the channel half, which you did not flag: `client.channels.fetch(id).catch(() => null)` collapsed a rate limit, a 5xx or a timeout into "channel no longer exists" and deleted the subscription. It now deletes only on a resolved `null` or an Unknown Channel/Guild code, and skips with a warning otherwise.

New `validate-data.test.ts` covers: a live install row is never orphaned (and costs zero REST calls), a Discord-confirmed removal is orphaned, a missing install row alone is not, and an unreachable Discord aborts without orphaning anything — including that it aborts on the first failure rather than accumulating a partial orphan list.

File: packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    httpSurface: "admin",
+    competitionActivityWorker: false,
+    databaseMetricSweeps: false,
+    databaseSeeding: false,
+  },
+  "activity-worker": {
+    championAssets: true,
+    voiceAssistant: false,
+    reportLakeFold: false,
+    temporalWorkers: DISCORD_WORKERS,
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts:231]: Real, and on investigation it is wider than `runReport` — I took your second remedy (give the worker the lake contract) rather than rerouting the dispatch.

I traced lake reachability per activity queue before choosing. **Every** embedded activity queue reads the lake, not just `background`: `realtime` reads it settling SQL dares (`postmatch → dare-settle-v3 → dare-sql-v3 → withDuckDBConnection`) and evaluating hall progression (`ingestMatch → progression/postmatch → evaluate-match → progression-lake-reads`); `interactive` reads it for every Explore query; `background` reads it from `runReport`, `runDetachedBackgroundWork` (parlay history/stats), `invokeScoutWeeklyParlayAction` and the `summoner-index-backfill` job; `lake` is the compactor. `realtime`, `background` and `lake` also write its staging directories, as does the competition worker.

So rerouting `runReport` to the lake queue would fix one symptom of several and would need `patched()` around it plus each of the others, while `realtime` and `interactive` have no reroute available at all — open-history cost for a partial fix. I changed no workflow code, so there is no replay or versioning story here.

Instead the capability table now declares the dependency and the boot sequence enforces it:
- new `reportLakeAccess` capability, separate from `reportLakeFold`. True for `combined`, `application` and `activity-worker`; false only for `gateway`, the one role that runs no activity queue.
- the `report-lake` boot step is emitted for any role with `reportLakeAccess`, and runs before `temporal-core` as well as before `http-server` — a role that folds publishes a build, a role that only reads asserts one exists and refuses to start otherwise. That is the specific fix for "recording empty results as successful": an unmounted lake is not an error for DuckDB, it scans zero parquet files and returns zero rows, so the pod must refuse rather than answer.
- tests assert that any role running a worker declares lake access, and that the lake is settled before both HTTP and the workers.

The consequence is worth stating plainly, and I have flagged it to the coordinator: `activity-worker` now needs the same volume `application` owns, and the cluster PVC is ReadWriteOnce, so the two cannot both mount it as things stand. That role is therefore not deployable until the lake becomes shareable (remote store, or every reader moved behind the `lake` queue). This PR ships roles only — Kubernetes still runs `combined` — so nothing regresses today, and the boot gate means a premature split fails loudly instead of silently serving empty reports. Documented in the backend README.

File: packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts:233]: Fixed — the competition-validation cleanup now asks an injected guild-validation port backed by the merged installed-guilds REST-confirmed lookups instead of client.guilds.cache: one bulk install-row query for the survey, a per-guild REST-confirmed check before anything destructive, and an unreachable Discord throws rather than guessing false. Correct on every role including combined; a new test proves a guild with an empty cache and a live install row is not orphaned.

File: packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts:231]: Fixed via honest capability modeling rather than a queue move (changing an activity's task queue inside workflow code is replay-breaking for open histories): reportLakeAccess — which every embedded activity queue turns out to need (realtime settles SQL dares and hall progression from the lake, interactive serves Explore, background runs reports and parlays) — is now declared separately from reportLakeFold (exactly one publisher), and a lake-reading role refuses to boot when the lake has no published build, since DuckDB scans an empty directory as zero rows and would record empty results as success. The Wave 6 deployment chooses between a shared read mount and queue redistribution, with this boot gate making the silent-empty topology impossible either way.

**Comment by shepherdjerred** [comment]
@codex review

<sub>Automated re-review request for `3b3facf` from the CI review gate.</sub>

**Comment by shepherdjerred** [comment]
@codex review

<sub>Automated re-review request for `b9149ad` from the CI review gate.</sub>

**Comment by shepherdjerred** [comment]
@codex review

<sub>Automated re-review request for `5c8ef39` from the CI review gate.</sub>

File: packages/scout-for-lol/packages/backend/src/consumer/access.ts [hunk 3]
```diff
@@ -53,17 +56,59 @@ export type ConsumerGuildAccessResult =
...
 ): Promise<ConsumerGuildAccessResult> {
   const guilds = await fetchUserGuildsForRequest(user);
   let connectedGuildIds: Iterable<string> | undefined;
   if (configuration.environment === "prod") {
-    const guildMembership =
-      await import("#src/discord/utils/guild-membership.ts");
-    connectedGuildIds = guildMembership.getConnectedServerIds();
+    connectedGuildIds = await installedGuildIdsOrUnavailable(
+      guilds.map((guild) => guild.id),
+      installedGuilds,
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/consumer/access.ts:140]: consumer/access.ts now confirms every guild that would grant: installedGuildIdsAmong narrows in one query, then each survivor goes through isScoutInstalledInGuild so a stale row cannot grant access to a server Scout has left, with the port's asymmetry preserved (live row + unreachable Discord still grants, with a warning); I confirm every candidate rather than short-circuiting on the first, because the returned list is also the Explore alias-resolution scope and truncating it would silently narrow which servers a caller can resolve names in — new confirmed-access.test.ts covers stale-row-denied, REST-down-still-grants, and no-row-never-asked.

File: packages/scout-for-lol/packages/backend/src/league/tasks/cleanup/validate-data.ts [hunk 4]
```diff
@@ -209,25 +266,31 @@ async function validateChannels(client: Client): Promise<void> {
...
       try {
-        // Try to fetch the channel
-        const channel = await client.channels
-          .fetch(channelId)
-          .catch(() => null);
-
-        if (!channel) {
-          // Channel doesn't exist
+        const channel = await client.channels.fetch(channelId);
+        if (channel === null) {
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/league/tasks/cleanup/validate-data.ts]: The channel half now reads through botRest().channel instead of client.channels.fetch — you were right that the helper resolves null for an un-cached guild after a successful REST read, so my earlier fix only covered the throw path and every subscribed channel would still have been deleted on activity-worker; the decision is extracted as resolveOrphanedChannelIds and tested for confirmed-absent, present, unreachable-orphans-nothing, and one-flaky-read-does-not-block-the-others.

File: packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+  application: {
+    championAssets: true,
+    voiceAssistant: false,
+    reportLakeAccess: true,
+    reportLakeFold: true,
+    temporalWorkers: ALWAYS_ON_WORKERS,
+    deferredTemporalWorkers: NO_WORKERS,
+    discordGateway: false,
+    gatewayReadyReconciliation: false,
+    httpSurface: "full",
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts:198]: Implemented as an honest refusal rather than a REST workaround, since Discord exposes no read for a member's current voice channel: a new voiceStateAccess capability is true exactly where discordGateway is, and customs voice moves throw CustomAuthHttpError(503) naming the gateway role — applied to arrangeCustomVoice, returnCustomVoiceToLobby AND cleanExpiredCustomVoice (the expiry sweep shares the same mover and would otherwise delete team channels out from under the players in them), guarded at the top of each entry point so nothing is committed first, with the Wave 6 routing options recorded in src/customs/voice-capability.ts.

File: packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts:233]: Both migrated to the install port — outreach asks isScoutInstalledInGuild per live install instead of client.guilds.cache.has (an empty cache was skipping every install as not_a_member, silently stopping outreach entirely), and the removed-guild reconciler now takes candidacy from installedGuildIdsAmong with the isReady() early-return deleted, which is safe because nomination only records a sighting and deletion still needs two Discord-confirmed 10004s a day apart; new tests cover both with a never-ready, empty-cache client.

File: packages/scout-for-lol/packages/backend/src/metrics/index.ts [hunk 6]
```diff
@@ -756,17 +757,23 @@ import "@scout-for-lol/backend/metrics/usage.ts";
...
   updateUptimeMetric();
-  await updateUsageMetrics();
-  await updateLimitMetrics();
-  await updateBettingMetrics();
-  const { updateScoutTemporalDurabilityMetrics } =
-    await import("#src/metrics/platform/temporal.ts");
-  await updateScoutTemporalDurabilityMetrics();
+  if (databaseMetricSweepsEnabled()) {
+    // Dynamic import to avoid circular dependency issues
+    const { updateUsageMetrics } = await import("./usage.js");
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/metrics/index.ts:769]: Removed the import-time updateUsageMetrics() call and the 5-minute interval from metrics/usage.ts so scheduling happens only through getMetrics(), which sweep-policy gates to the owning role — the gauges now refresh per scrape instead of per timer, and I left a comment (and am flagging here) that the durable home for the recurring sweep is a Temporal Schedule, deliberately not migrated in this PR because that changes who writes the gauges.

File: packages/scout-for-lol/packages/backend/src/consumer/access.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/consumer/access.ts:140]: Fixed — the guilds that would grant consumer access now run the REST-confirmed isScoutInstalledInGuild check (short-circuiting on the first confirmed guild, bounded by its 60s cache) with the established asymmetry: a confirmed-absent guild does not grant, an unreachable Discord trusts the live row with a warning. Tested for both the stale-row-denied and REST-down-trusted paths.

File: packages/scout-for-lol/packages/backend/src/league/tasks/cleanup/validate-data.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/league/tasks/cleanup/validate-data.ts]: Fixed — the channel half now reads through the raw bot-REST channel lookup, which is cache-independent: only a confirmed not-found counts as absence, an unreachable Discord throws rather than orphaning, and the empty-cache test now covers the channel half alongside the guild half.

File: packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts:198]: Fixed as an honest refusal: Discord exposes no endpoint to READ a member's current voice channel (only to set it), so REST cannot supply what the move decision needs. The capability table now declares voiceStateAccess (gateway/combined only) and every arrangement entry point refuses loudly up front on roles without it — 503 with an operator-facing message — instead of moving nobody and reporting ready; per-role refusal is tested, and routing gateway-coupled mutations to the gateway role via a dedicated Temporal queue is recorded in the module docblock as the Wave 6 mechanism.

File: packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/configuration/runtime-role.ts:233]: Fixed — outreach and removed-guild reconciliation now use the installed-guilds port instead of guilds.cache, and the isReady() early-return became a check on the data source the reconciliation actually needs, so removals missed during gateway downtime reconcile on any role. The refactor simplified reconcile-removed-guilds enough to drop one cognitive-complexity suppression. Both paths tested with an empty-cache client.

File: packages/scout-for-lol/packages/backend/src/metrics/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/metrics/index.ts:769]: Fixed — metrics/usage.ts no longer runs a sweep at import or installs its own interval; collection happens only through the owner-gated scrape path, so databaseMetricSweeps is now a real single-owner contract, and the module comment records that the durable home for a recurring sweep is a Temporal Schedule (flagged as follow-up since a Schedule changes who writes the gauges).

**Comment by shepherdjerred** [comment]
@codex review

<sub>Automated re-review request for `0df43c2` from the CI review gate.</sub>

File: packages/scout-for-lol/packages/backend/src/discord/rest-identity.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * Put the bot token on the client's REST manager.
+ *
+ * Safe on every role: `login()` sets the same token again, so a gateway-owning
+ * process is unaffected by having had it set earlier. Deliberately not done at
+ * import time — the module graph is loaded by tests and scripts that have no
+ * real token, and this is a decision the composition root makes once.
+ */
+export function authorizeDiscordRest(target: Client = client): void {
+  target.rest.setToken(configuration.discordToken);
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/discord/rest-identity.ts:33]: Fixed — a new fetchChannelForDelivery helper resolves channels with allowUnknownGuild: true and every delivery path from the sends audit now uses it (sender + failure diagnosis, dare callouts, parlay publish/refresh, weekly refresh, message refresh, customs recruitment); the diagnosis path was equally cache-poisoned and now uses a three-state permission check so a gatewayless role never reports 'channel deleted' for a live channel or DMs an owner about a permission guess — genuinely labeled 50013/50001 errors still classify from the error itself. Tested: an empty-cache client resolves and delivers, a true 10003 still classifies as missing.

**Comment by shepherdjerred** [comment]
@codex review

<sub>Automated re-review request for `e99db38` from the CI review gate.</sub>

**Comment by shepherdjerred** [comment]
### Review findings dismissed by an operator

Each entry was marked resolved in the review comment after being verified fixed at this head or incorrect. The gate honours the resolved state, so this comment is the record of why.

- **thread PRRT_kwDOHf4r4c6g8zJP** — Fixed at head e99db38d41: fetchChannelForDelivery resolves with allowUnknownGuild:true and every delivery path from the sends audit uses it (sender + failure diagnosis, dare callouts, parlay publish/refresh, weekly refresh, message refresh, customs recruitment); the diagnosis path now uses a three-state permission check so a gatewayless role neither reports live channels as deleted nor guesses permission revocations, while labeled 50013/50001 errors still classify from the error. Tested with an empty-cache client end-to-end. Codex re-listed the thread without striking after the fix was pushed.
